src/sinks/elasticsearch.rs

```diff
@@ -263,14 +282,45 @@ impl RetryLogic for ElasticSearchRetryLogic {
             _ if status.is_server_error() => RetryAction::Retry(
                 format!("{}: {}", status, String::from_utf8_lossy(response.body())).into(),
             ),
+            _ if status.is_client_error() => {
+                let body = String::from_utf8_lossy(response.body());
+                warn!(
+                    message = "client error",
+                    body = %body,
+                    rate_limit_secs = 30
+                );
+                RetryAction::DontRetry("client error".into())
+            }
             _ if status.is_success() => {
                 let body = String::from_utf8_lossy(response.body());
                 match body.find("\"errors\":true") {
-                    Some(_) => match serde_json::from_str::<Value>(&body) {
-                        Err(_) => RetryAction::DontRetry(
-                            "some messages failed, and invalid response from elasticsearch".into(),
-                        ),
-                        Ok(_data) => RetryAction::DontRetry("some messages failed".into()),
+                    Some(_) => match serde_json::from_str::<ESResultResponse>(&body) {
+                        Err(json_error) => {
+                            warn!(
+                                message = "ElasticSearch unparsable error response",
+                                %json_error,
+                                rate_limit_secs = 30
+                            );
+                            RetryAction::DontRetry(
+                                "some messages failed, and invalid response from elasticsearch"
+                                    .into(),
+                            )
+                        }
+                        Ok(esrr) => {
+                            for esri in esrr.items {
+                                if let Some(error_obj) = esri.index.error {
+                                    warn!(
+                                        message = "ElasticSearch error response",
+                                        err_type = %error_obj.err_type,
+                                        reason = %error_obj.reason,
+                                        rate_limit_secs = 30
+                                    );
+                                    // Rate limiting would ignore any others...
+                                    break;
+                                }
+                            }
```

Comment:
Let Rust's iterator adapters help you out here:

```suggestion
                            esrr.items
                                .into_iter()
                                .filter_map(|item| item.index.error)
                                .next()
                                .and_then(|error| warn!(
                                    message = "ElasticSearch error response",
                                    err_type = %error.err_type,
                                    reason = %error.reason,
                                    rate_limit_secs = 30
                                ))
```